fix(group): set default group chat title when provided title is invalid

Kribylet opened the pull request:
Add handling logic for when a group exists but has an invalid title.
Title is created using the first 8 hexadecimals of the group chat id.

Fixes #5542 and #5694

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5695)
<!-- Reviewable:end -->

Kribylet committed:
fix(group): set default group chat title when provided title is invalid

Add handling logic for when a group exists but has an invalid title.
Title is created using the first 8 hexadecimals of the group chat id.

sudden6 reviewed (changes requested):
Please reformat according to https://github.com/qTox/qTox/blob/master/doc/coding_standards.md, you can use the script linked in this file.

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5695#-:-LhJX4hN0UfnOtrDlQey:bw9s30i)* status: 0 of 1 LGTMs obtained

___
*[src/core/core.cpp, line 1036 at r1](https://reviewable.io/reviews/qtox/qtox/5695#-LhJXbYX7uYRFJFrNz_k:-LhJXbYX7uYRFJFrNz_l:b-8l3t9w) ([raw file](https://github.com/qtox/qtox/blob/a7d1a5189f399a72bfda50dfbfb82f1022212c5e/src/core/core.cpp#L1036)):*
> ```cpp
>         QByteArray nameByteArray;
>         QString name;
>         bool invalidTitle;
> ```

directly initialize this variable with the function call a few lines below?
___
*[src/core/core.cpp, line 1040 at r1](https://reviewable.io/reviews/qtox/qtox/5695#-LhJY4LUAmdkTcBdi_h2:-LhJY4LUAmdkTcBdi_h3:bavlasm) ([raw file](https://github.com/qtox/qtox/blob/a7d1a5189f399a72bfda50dfbfb82f1022212c5e/src/core/core.cpp#L1040)):*
> ```cpp
>         size_t titleSize = tox_conference_get_title_size(tox.get(), groupNumber, &error);
>         invalidTitle = LogConferenceTitleError(error);
>         if (!invalidTitle)
> ```

merge this with the `if (invalidTitle)`and use local variables to make it easier to read?


<!-- Sent from Reviewable.io -->

Kribylet commented:
> > ```c++
> >         QByteArray nameByteArray;
> >         QString name;
> >         bool invalidTitle;
> > ```
> 
> directly initialize this variable with the function call a few lines below?
> 
> >         size_t titleSize = tox_conference_get_title_size(tox.get(), groupNumber, &error);
> >         invalidTitle = LogConferenceTitleError(error);
> >         if (!invalidTitle)
> > ```
> 
> merge this with the `if (invalidTitle)`and use local variables to make it easier to read?

I will reformat. I really tried to condense the boolean logic, but if we entertain the possibility that we can get an error in between retrieving the title size and the title name, there will either be code duplication or this kind of awkward bool variable. I'm new to the qTox project so if you tell me which you consider the lesser evil I will adjust accordingly!

sudden6 reviewed (changes requested):
I just now realized that you re-use the `invalidTitle` variable for the second error check. I think you are correct in that there can be errors between those two function calls, but it may be enough to guard against that using an `assert(error == TOX_ERR_CONFERENCE_TITLE_OK)`.

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5695#-:-LhMZcC_9YiqHUq2kJHk:buj8prt)* status: 0 of 1 LGTMs obtained

___
*[src/core/core.cpp, line 1046 at r1](https://reviewable.io/reviews/qtox/qtox/5695#-LhMZAkBBvcDwZLwmcbT:-LhMZAkBBvcDwZLwmcbU:b-5jfby4) ([raw file](https://github.com/qtox/qtox/blob/a7d1a5189f399a72bfda50dfbfb82f1022212c5e/src/core/core.cpp#L1046)):*
> ```cpp
>             invalidTitle = LogConferenceTitleError(error);
>         }
>         if (error == TOX_ERR_CONFERENCE_TITLE_CONFERENCE_NOT_FOUND)
> ```

after further reviewing, I'm not 100% sure what you intend to do with this line. Why should all error cases, besides `TOX_ERR_CONFERENCE_TITLE_CONFERENCE_NOT_FOUND` be ignored? How could this error code even come up, when we iterate over the groups that c-toxcore provides to us? Did you observer that? (Then it's probably also a c-toxcore bug :) )


<!-- Sent from Reviewable.io -->

Kribylet commented:
When I wrote the code I studied the toxcore functions being called, and they both have three possible outcomes for "error":
TOX_ERR_CONFERENCE_TITLE_INVALID_LENGTH,
TOX_ERR_CONFERENCE_TITLE_CONFERENCE_NOT_FOUND and
TOX_ERR_CONFERENCE_TITLE_OK

It's possible that we could never run into CONFERENCE_NOT_FOUND. My reasoning was that the previous code did react to this possibility, and that my change should attempt to respect the previous behavior outside of my changes. If a title has invalid length, we simply substitute our own title, so all cases are covered at that point. (Invalid title is assigned a title by us, a valid title is retrieved from tox)

And finally to argue my preference, I think we should respect the possibility of a group ceasing to exist after we've retrieved the group list on the basis that qTox is multi-threaded. My understanding of what is running concurrently isn't perfect, but if we consider it possible that we can retrieve a valid title size and then get an invalid title length error in the following lines of code, it doesn't seem far-fetched with the same reasoning to suppose a group could be closed, too.

sudden6 commented:
> I think we should respect the possibility of a group ceasing to exist after we've retrieved the group list on the basis that qTox is multi-threaded. My understanding of what is running concurrently isn't perfect, but if we consider it possible that we can retrieve a valid title size and then get an invalid title length error in the following lines of code, it doesn't seem far-fetched with the same reasoning to suppose a group could be closed, too.

Multi-threading will not be an issue, since in https://github.com/qTox/qTox/pull/5695/files#diff-5db0ca19c12b31af1e97555eeadc9eeaR1022 we acquire a lock that (should) guard against that. So there should be no creations/destructions of groups possible. If it still happens c-toxcore is breaking some promises and we have bigger problems ;)

Kribylet committed:
fix(group): Condense invalid title handling logic, clang-format

This commit ignores the possibility of invalid length error being thrown
by a title after we've already checked its size the first time. We also
assume that a group chat cannot cease existing while we iterate over the
retrieved groups.

Kribylet commented:
Sorry about the messy pushes, I retracted the commit when I saw I had missed condensing the QByteArray initialization.

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5695#-:-Lhg-WK08OXhxFBJAtWO:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained

___
*[src/core/core.cpp, line 1036 at r1](https://reviewable.io/reviews/qtox/qtox/5695#-LhJXbYX7uYRFJFrNz_k:-Lhg-J4L6ZWeiX8k8P3p:b32wtd) ([raw file](https://github.com/qtox/qtox/blob/a7d1a5189f399a72bfda50dfbfb82f1022212c5e/src/core/core.cpp#L1036)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

directly initialize this variable with the function call a few lines below?
</blockquote></details>

ok
___
*[src/core/core.cpp, line 1040 at r1](https://reviewable.io/reviews/qtox/qtox/5695#-LhJY4LUAmdkTcBdi_h2:-Lhg-SoR1-esqCAnyrgQ:b32wtd) ([raw file](https://github.com/qtox/qtox/blob/a7d1a5189f399a72bfda50dfbfb82f1022212c5e/src/core/core.cpp#L1040)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

merge this with the `if (invalidTitle)`and use local variables to make it easier to read?
</blockquote></details>

ok
___
*[src/core/core.cpp, line 1046 at r1](https://reviewable.io/reviews/qtox/qtox/5695#-LhMZAkBBvcDwZLwmcbT:-Lhg-TjZ4VZAPrWe3loW:b32wtd) ([raw file](https://github.com/qtox/qtox/blob/a7d1a5189f399a72bfda50dfbfb82f1022212c5e/src/core/core.cpp#L1046)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

after further reviewing, I'm not 100% sure what you intend to do with this line. Why should all error cases, besides `TOX_ERR_CONFERENCE_TITLE_CONFERENCE_NOT_FOUND` be ignored? How could this error code even come up, when we iterate over the groups that c-toxcore provides to us? Did you observer that? (Then it's probably also a c-toxcore bug :) )
</blockquote></details>

ok


<!-- Sent from Reviewable.io -->
